### DISCUSSION TITLE
Problems running CROCO output in OpenDrift

### DISCUSSION THREAD

**Discussion by N1coleR** [OpenDrift/opendrift#1367] [Q&A] [answered]
Dear OpenDrift community:

I am experiencing some problems running an output from Croco in OpenDrift.

I have merged the croco_grd.nc and the variable Vstretching (from the .bry file) to my croco_his.nc file; it should work that way. But during the run, I get some warnings and errors like:
01:38:40 WARNING opendrift.readers.reader_ROMS_native:183: Vtransform not found, using 1

01:38:40 WARNING opendrift.readers.basereader.structured:50: No proj string or projection could be derived, using 'fakeproj'. This assumes that the variables are structured and gridded approximately equidistantly on the surface (i.e. in meters). This must be guaranteed by the user. You can get rid of this warning by supplying a valid projection to the reader.

01:38:46 WARNING opendrift.models.basemodel:730: Seafloor check not being run because environment is missing. This will happen the first time the function is run but if it happens subsequently there is probably a problem.

01:38:46 ERROR   opendrift.models.basemodel.environment:666: No mask found for x_sea_water_velocity

At the end of the simulation, my particles do not move. I recently shared my file with another colleague with a different operating system (Linux), and the file works fine for him. My operating system is Windows. I reinstalled Anaconda and OpenDrift, but they still do not work for me (the examples that are in OpenDrift work fine, but my Croco file does not). Has anyone experienced something similar?

I'd really appreciate it if anyone could look at my file and code (.txt).

Thank you in advance.
Nicole Reguera

[OpenDrift_Test.txt](https://github.com/user-attachments/files/16338552/OpenDrift_Test.txt)

Link to croco_var.nc https://drive.google.com/file/d/1vnoP8uW6rhDwx93cWGG2JLj1BHRsPoRu/view?usp=drive_link

**Comment by knutfrode** [comment]
Hi,

The listed warnings are not a problem here, but there seem to be a mismatch for the mask of velocity component (variable u -> x_sea_water_velocity)

In your file (that I downloaded), the dimensions of u seems to be
`float u(time, s_rho, eta_rho, xi_u)`
whereas for a ROMS file that I have available (and which works fine), the corresponding dimensions are
`float u(ocean_time, s_rho, eta_u, xi_u)`

Thus first horizontal spatial coordinate in your Croco file is `eta_rho`, whereas it is `eta_u` for the ROMS file:
I am not sure whether this is a deliberate difference between Croco and ROMS, or whether this is a bug either in Croco, or caused by your merging?

I am not an expert (nor a user) of ROMS native files, but perhaps @kthyng knows more about the difference between ROMS and Croco?

**Comment by kthyng** [comment]
Yes I have the same dimensions for `u` as @knutfrode:

```python
 ds.u
Out[5]: 
<xarray.DataArray 'u' (ocean_time: 210249, s_rho: 30, eta_u: 1044, xi_u: 723)>
dask.array<open_dataset-u, shape=(210249, 30, 1044, 723), dtype=float32, chunksize=(1, 15, 522, 362), chunktype=numpy.ndarray>
Coordinates:
    lat_u       (eta_u, xi_u) float64 dask.array<chunksize=(1044, 723), meta=np.ndarray>
    lon_u       (eta_u, xi_u) float64 dask.array<chunksize=(1044, 723), meta=np.ndarray>
  * ocean_time  (ocean_time) datetime64[ns] 1999-01-01T01:00:00 ... 2023-01-01
  * s_rho       (s_rho) float64 -0.9833 -0.95 -0.9167 ... -0.05 -0.01667
Dimensions without coordinates: eta_u, xi_u
Attributes:
    cell_methods:   ocean_time: point
    field:          u-velocity
    grid:           grid
    location:       edge1
    long_name:      u-momentum component
    standard_name:  sea_water_x_velocity
    time:           ocean_time
    units:          meter second-1
```

However, the main things are that the u_mask exist and that the dimensions match the variable identified as `x_sea_water_velocity`

https://github.com/OpenDrift/opendrift/blob/8022e94daecaae00d3a68858605736925e061739/opendrift/readers/reader_ROMS_native.py#L561-L568

I don't have access to that file to look at the variables myself, and I am not familiar with CROCO, just ROMS.

**Comment by N1coleR** [comment]
Hi,

Sorry for the delay, I was on vacation.

The original CROCO file (without the merge) has the following dimensions for u and v

    u         
           Size:       435x468x32x248
           Dimensions: xi_u,eta_rho,s_rho,time
           Datatype:   single
           Attributes:
                       long_name     = 'u-momentum component'
                       units         = 'meter second-1'
                       field         = 'u-velocity, scalar, series'
                       standard_name = 'sea_water_x_velocity_at_u_location'
                       coordinates   = 'lat_u lon_u'
    v         
           Size:       436x467x32x248
           Dimensions: xi_rho,eta_v,s_rho,time
           Datatype:   single
           Attributes:
                       long_name     = 'v-momentum component'
                       units         = 'meter second-1'
                       field         = 'v-velocity, scalar, series'
                       standard_name = 'sea_water_y_velocity_at_v_location'
                       coordinates   = 'lat_v lon_v'


I still do not know why my file does not work in OpenDrift

**Reply by knutfrode** [reply]
There seem to be a mismatch between the current velocity variables and the available masks.
E.g. the variable `u` has dimensions `time, s_rho, eta_rho, xi_u`

But there is no mask with corresponding horizontal dimensions `eta_rho, xi_u`, the only available masks are:
```
	double mask_psi(eta_psi, xi_psi) ;
		mask_psi:long_name = "mask on PSI-points" ;
		mask_psi:option_0 = "land" ;
		mask_psi:option_1 = "water" ;
	double mask_rho(eta_rho, xi_rho) ;
		mask_rho:long_name = "mask on RHO-points" ;
		mask_rho:option_0 = "land" ;
		mask_rho:option_1 = "water" ;
	double mask_u(eta_u, xi_u) ;
		mask_u:long_name = "mask on U-points" ;
		mask_u:option_0 = "land" ;
		mask_u:option_1 = "water" ;
	double mask_v(eta_v, xi_v) ;
		mask_v:long_name = "mask on V-points" ;
		mask_v:option_0 = "land" ;
		mask_v:option_1 = "water" ;
```

It also sounds unlogical that the dimensions for `u` should be `eta_rho, xi_u`.
Logical should be as for ROMS: `eta_u, xi_u`

Thus perhaps a bug in this version of Croco?

**Comment by knutfrode** [comment]
Btw, I noticed that there are several more errors in this croco-file, e.g. the standard_name of current variables are of the form `sea_water_x_velocity_at_u_location`, where correct should be `sea_water_x_velocity` (See table of standard_names: https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html)

So as a user of Croco, perhaps you could report to the developers that their output (in particular standard_names) is not CF-compliant, and also mentioned the bug above that wrong dimensions are used for current variables?

**Comment by N1coleR** [comment]
Hi @knutfrode 

I am going to check that points.

Thank you very much!

**Comment by kthyng** [comment, marked as answer]
Really, `eta_rho` and `eta_u` are the same values and `xi_rho` and `xi_v` are the same values. You can see this in this ROMS grid image if you note that eta is in the j index direction and xi is the i index direction: https://www.myroms.org/wiki/Grid_Generation. So that is why they may be mixed up. However, the ROMS reader would need to be written more generally to accept that. Shouldn't be too hard but it might be easier at least initially to first change the names in your file and see if that fixes the problem? I won't be able to do this change in the ROMS reader right away myself if it is the issue

I can't think of why it would work on a linux machine though, if this is the issue?

**Comment by N1coleR** [comment]
@kthyng thank you very much. Changing the names works just fine!

**Reply by N1coleR** [reply]
@knutfrode Also I change the standard names, thank you!!!

**Reply by Sabarrera19** [reply]
Hello, I have the same problem. Could you tell me which names you changed and in which variables to get the model running? Thank you very much.

**Reply by N1coleR** [reply]
Hello @Sabarrera19.

First I merged the croco_grd file and the variable Vstretching (from the bry file) to my croco_his file. 
Then you have to change the dimension and the attribute names from the U and V variables, to make them match with mask_u and mask_v attributes (from the croco_grd file):

U and V dimension names should be:
U --> xi_u, eta_u, x_rho, time   instead of xi_u, eta_rho, etc.
V --> xi_v, eta_v, x_rho, time    instead of xi_rho, eta_v, etc.

and the attribute names should be:

for U --> sea_water_x_velocity 
for V --> sea_water_y_velocity


The NCO code:

1) Extract U and V variables

ncks -v u croco_his.nc u.nc
ncks -v v croco_his.nc v.nc

2) Change the dimensions to make them match with mask_u y mask_v

ncrename -d xi_rho,xi_v v.nc
ncrename -d eta_rho,eta_u u.nc

3) Change the attribute names

ncatted -O -a standard_name,v,o,c,"sea_water_y_velocity" v.nc
ncatted -O -a standard_name,u,o,c,"sea_water_x_velocity" u.nc


4) From the original .nc (croco_his.nc) eliminate u and v variables 

ncks -x -v u,v croco_his.nc croco_his2.nc

5) Add the changed u and v variables

ncks -A u.nc croco_his2.nc
ncks -A v.nc croco_his2.nc


You also have to fix the time variable, the following link could help you:
https://github.com/OpenDrift/opendrift/discussions/1335

**Reply by knutfrode** [reply]
Note that the merging of the gridfile might be unnecessary, as the ROMS reader also can take a gridfile as input in addition to the data files/URLs:
`r = reader_ROMS_native.Reader(<netCDF-files>, gridfile=<gridfile>)`
https://github.com/OpenDrift/opendrift/blob/master/opendrift/readers/reader_ROMS_native.py#L78

**Comment by knutfrode** [comment]
Very good!
In principle we could update the ROMS reader to handle this case. On the other hand, to my understanding this is still a formal error (non--compliance with CF-convention) in the output from Croco.

Thus I believe it is better in the long run to rather correct such errors at their origin (in this case Croco), instead of making patches in downstream softwares (in this case OpenDrift).
